eBPF helper function for attribute search in the netlink message

Kalimuthu-Velappan opened the pull request:
This patch adds the support for following helper function.
    FN(skb_get_nlattr),
    FN(skb_get_nlattr_nest)

skb_get_nlattr:
    Find a specific attribute in a stream of attributes

skb_get_nlattr_nest:
    Find a specific attribute in a stream of nested attributes

lguohan commented:
is this upstream patch? if not, can you upstream this patch first?

paulmenzel commented on patch/netlink-socket-attribute-filter.patch:
```diff
@@ -0,0 +1,70 @@
+diff --git a/include/uapi/linux/bpf.h b/include/uapi/linux/bpf.h
```
Please use a `git format-patch` formatted patch, which can be easily applied with `git am ….patch` to the Linux kernel source. Please add a descriptive commit summary, message and Signed-off-by line.

Kalimuthu-Velappan replied:
Fixed

paulmenzel commented on patch/netlink-socket-attribute-filter.patch:
```diff
@@ -0,0 +1,70 @@
++	.ret_type	= RET_INTEGER,
++	.arg1_type	= ARG_PTR_TO_CTX,
++	.arg2_type  = ARG_ANYTHING,
++	.arg3_type  = ARG_ANYTHING,
```
Please align the equal signs with tabs as is done in the rest of the file.

Kalimuthu-Velappan replied:
Fixed

paulmenzel commented on patch/netlink-socket-attribute-filter.patch:
```diff
@@ -0,0 +1,70 @@
++	.ret_type	= RET_INTEGER,
++	.arg1_type	= ARG_PTR_TO_CTX,
++	.arg2_type  = ARG_ANYTHING,
++	.arg3_type  = ARG_ANYTHING,
```
Ditto.

Kalimuthu-Velappan replied:
Fixed

paulmenzel reviewed (commented):
> is this upstream patch? if not, can you upstream this patch first?

I couldn’t find it in Linus’ Linux master branch v5.10-rc6-178-gb3298500b23f. So yes, please send it upstream first.

Can you also give a small example, how to use the new helper functions? Please also not on what device you tested this.

Kalimuthu-Velappan commented:
> is this upstream patch? if not, can you upstream this patch first?

We have already upstreamed this to the Linux community. But it is taking time to merge.
https://www.spinics.net/lists/netdev/msg627990.html

I will address your comments and resubmit the patch agagin.

lguohan commented:
i looked the thread, the question is reasonable, we will need to wait till that one gets signed off.

Kalimuthu-Velappan commented:
Fixed the alignment issue and added the descriptive summary. 

The PR 6131 provides the library support on how to use these attributes.
https://github.com/Azure/sonic-buildimage/pull/6131

ben-gale commented:
@Kalimuthu-Velappan , @lguohan , @paulmenzel - this one has been quiet for a few days - can we get it moving towards merge pls?

@lguohan - I'm not sure I understand your comment saying "i looked the thread, the question is reasonable, we will need to wait till that one gets signed off" - which thread and which question are you talking about please?

paulmenzel commented:
Could you please rebase it on the master branch? Currently already committed changes are present in the commit list.

Kalimuthu-Velappan commented:
retest this please

jarias-lfx commented:
/easycla
